### ISSUE TITLE
Method Callbacks not being called intermittently

### ISSUE THREAD

**Issue by ppotoplyak** [meteor/meteor#7868]
In the following the client calls server method `recordPing(counter)` every 3 secs, with `counter` being incremented by one in the method callback.

Client:

```
  var pingCount = 0;
  function heartbeat() {
    console.log('->recordPing pingCount: ' + pingCount);
    Meteor.call('recordPing', pingCount, function(err) {
      if(!angular.isUndefined(err)) alert(err);
      console.log('<-recordPing pingCount: ' + pingCount);
      pingCount = pingCount + 1;
    });
    heartbeatTimeout = setTimeout(heartbeat, 3000);
  }
  heartbeat();
```

Server:

```
Meteor.methods({
  recordPing: function(pingCounter) {
    console.log('->recordPing pingCounter : ' + pingCounter + 
               (this.connection != null ? ', called from client' : ''));
  }
});
```

On Safari/iPhone5 I am intermittently seeing multiple `recordPing` invocations with the _same_ counter value.  This could only happen if the method return callback was being delayed.

Here is the server log showing a period where the counter is not incremented:
![serverlog](https://cloud.githubusercontent.com/assets/4952152/19088044/09b51930-8a3a-11e6-80d7-752754b993fe.png)

The client console confirms that the callback was delayed for 80+ seconds.  The client made a call to `pingCounter(111)` at 13:07:25 and executed a callback at 13:08:46.

![clientdelay](https://cloud.githubusercontent.com/assets/4952152/19088819/8cd061b4-8a3d-11e6-915a-33016b78531b.png)

Setup info:

```
ppotoplyak@bigscreen:/pit/PixelUprise2$ meteor --version
Meteor 1.4.1.1
ppotoplyak@bigscreen:/pit/PixelUprise2$ meteor list
angular-with-blaze                       1.3.11  Everything you need to use both AngularJS and Blaze templates in your Meteor app
angular:angular-cookies                  1.5.3_1  AngularJS (official) release. For full solution: http://angular-meteor.com/
angular:angular-sanitize                 1.5.3_1  AngularJS (official) release. For full solution: http://angular-meteor.com/
angularui:angular-ui-bootstrap           0.13.0  Native AngularJS (Angular) directives for Bootstrap.
angularui:angular-ui-router              0.2.15  angular-ui-router (official): Flexible routing with nested views in AngularJS
chrismbeckett:toastr                     2.1.2_1  Gnome / Growl type non-blocking notifications
cordova:cordova-google-play-services     25.0.0
cordova:cordova-plugin-device            1.1.1
cordova:cordova-plugin-underdevelopment  file://.cordova-sv-plugin/
cordova:cordova-plugin-vibration         2.1.1
es5-shim                                 4.6.14_1  Shims and polyfills to improve ECMAScript 5 support
fourseven:scss                           3.10.0  Style with attitude. Sass and SCSS support for Meteor.js.
http                                     1.2.9_1  Make HTTP calls to remote servers
jquery                                   1.11.9  Manipulate the DOM using CSS selectors
materialize:materialize                  0.97.6* Materialize (official): A modern responsive front-end framework based on Material Design
meteor-base                              1.0.4  Packages that every Meteor app needs
mobile-experience                        1.0.4  Packages for a great mobile user experience
mongo                                    1.1.12_1  Adaptor for using MongoDB and Minimongo over DDP
mrt:modernizr-meteor                     2.6.2  Modernizr repackaged for Meteor
natestrauser:animate-css                 3.5.1  Animate.css packaged for meteor
session                                  1.1.6  Session variable
shell-server                             0.2.1  Server-side component of the `meteor shell` command.
standard-minifier-css                    1.2.1  Standard css minifier used with Meteor apps by default.
standard-minifier-js                     1.2.0_1  Standard javascript minifiers used with Meteor apps by default.
stevermeister:angular-flipclock          0.1.1_4  AngularJS wrapper for FlipClock.js
tracker                                  1.1.0  Dependency tracker to allow reactive callbacks
twbs:bootstrap                           3.3.6  The most popular front-end framework for developing responsive, mobile first projects on the web.
```

I tried [SO](http://stackoverflow.com/questions/39709875/how-to-root-cause-observechanges-latency-spikes) & [Meteor Forums](https://forums.meteor.com/t/does-meteor-have-a-make-check-or-some-other-long-running-tests/29920) while reducing the issue.

**Comment by ppotoplyak** [comment]
The issue is actually worse.  It's not that the callbacks are being delayed, but they are completely dropped. i.e. callbacks are sporadically never ran, I see no errors on the client or the server.

I know you guys are going to ask for a repro which will be difficult given the intermittent nature and the larger context of the project.

Are there any known contexts in which Method callbacks are silently ignored and never executed?  The Methods are for sure executing, it's client the callbacks which are sporadically ignored.

**Comment by ppotoplyak** [comment]
If I use [Meteor.apply](http://docs.meteor.com/api/methods.html#Meteor-apply) and place the ```pingCount``` increment in ```onResultReceived``` then the server will stop seeing ```recordPing``` calls with the same ```pingCount```.  i.e. no callbacks are "lost"

This might be an issue where the local cache is in some odd state during which normal ```asyncCallback``` are not being triggered? Does this sound plausible?

**Comment by dard12** [comment]
We seem to be having the same issue where after a `Meteor.call`, the server run of the method quickly runs to completion but the callback is intermittently delayed many seconds or dropped entirely.

We also noticed that the callback can be dropped even when Minimongo is already reflecting writes from the method. This makes me wonder what the "local cache" for [Meteor.apply](http://docs.meteor.com/api/methods.html#Meteor-apply) refers to. Knowing what else it's waiting for could help diagnose what's happening.

**Comment by andylash** [comment]
I had the same problem.  I worked around it in one case by using onResultReceived rather than the regular callback, which works ok for my case (no need to wait for minimongo update).

**Comment by andylash** [comment]
Same problem with 1.4.2.3.  I had to rollback once I realized it was happening elsewhere.  Will need to debug this more fully

**Comment by abernix** [comment]
Thank you for filing the issue.

Unfortunately, I'm not seeing the problem that you're describing and I believe I've created an accurate reproduction of this issue [in this repo].(https://github.com/abernix/meteor-issue-7868) (Also, in the future, _please_ understand that submitting issues with reproductions already made will make things more efficient!)

I've let it run for two hours from Safari with no issues.

Are you saying that it is _only_ happening on Safari?  Or only on Safari on iPhone 5?  If it's only on an iPhone, what version of iOS is this occurring on?  Have you been able to reproduce on other devices?

More information is necessary and help debugging is appreciated (thanks, @andylash).

@dard12 I'm not sure your issue is the same.  But by the sounds of it, you should probably check out the [`onResultReceived` parameter](http://docs.meteor.com/api/methods.html#Meteor-apply) to debug if the delay you're experiencing is in updating the local cache (which I believe in this context is simply the local Mongo database collection).  If you are doing large updates which are updating a lot of documents, this could be your issue.  Either way, please open a new issue with your reproduction unless you're sure this is the same issue.

**Comment by ppotoplyak** [comment]
@abernix thank you for looking into this and the Meteor project in general.

The code snippets provided run a larger context with several subscriptions and data operations occurring, but nothing crazy. i.e. I am not surprised that the snippets alone are not enough to repro.

I think @dard12 has a point in that knowing what is being waited on (before firing the callback) could help in root causing.

I would love to give you a repro but I am not sure how easy that would be given the project size and the intermittent nature.  We spent 60+ hours reducing the problem which originally manifasted as ``observeChanges`` callbacks getting lost.  Spent 60+ more hours getting rid of subscriptions, changing everything to use ``Meteor.apply``, and replacing ``observeChanges`` with polling code.

Alternatively, I could give you private repo access and directions.

The following is anecdotal.  Issue occurred across several permutations of Android and iOS devices and browsers.  The rate of occurrence was correlated to the device horsepower and carrier type.   The weaker the device and carrier brand, the higher the rate of occurrence. e.g.numerous Galaxy S5s on Verizon _never_ saw the issue.  iPhone 5 on AT&T pay as you go sim would hit the issue ~once an hour.  Sprint seemed susceptible as well.

It would be great to see the [post about a torture](https://forums.meteor.com/t/does-meteor-have-a-make-check-or-some-other-long-running-tests/29920) test answered.  Perhaps the issue could exposed through a simpler angle.

**Comment by andylash** [comment]
The problems I'm seeing have all been on chrome desktop 55.  It could well occur on other browsers, but that's where i've seen it  Just mentioning it as the other posts were focussed on mobile, which this might not relate to.

**Comment by abernix** [comment]
@ppotoplyak I believe that waiting for the Mongo database collection on the client is the only thing which is delayed before firing the callback.   It could also be caused by re-synchronization if the connection to the server has to be re-established.  As far as I'm aware, all of the logic for when that callback would be fired is in [this file](https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js) and should be handled by `_runAfterUpdateCallbacks`.

You can also look at the `_runWhenAllServerDocsAreFlushed` calls to see anything which would add to that queue (which appears to only be added to when `updated` or `ready` messages are on the DDP wire).

I might recommend using a combination of `onResultReceived` and (visually) observing what is happening on the `websocket` endpoint by observing the "Frames" in Chrome DevTools under "Network":

![image](https://cloud.githubusercontent.com/assets/841294/20485530/a8b7775e-b004-11e6-9d49-1a6e2242fb98.png)

@andylash Thanks for reporting that it's happening on Desktop too.  Always easier to debug knowing the easiest way to reproduce it.

**Comment by ppotoplyak** [comment]
@abernix my first thought was connection issues and thus this [SO post](http://stackoverflow.com/questions/39559730/what-is-the-behavior-of-observechanges-when-the-internet-connection-is-temporari).  Then I realized that the client console is clean and that client method calls were occurring just fine, it was only the callbacks that were not firing.

Thanks for the debug pointers, I will spend some time on this next week.

**Comment by ppotoplyak** [comment]
@abernix @andylash I am able to repro using ```Meteor 1.4.2.3```

The setup is Safari on iOS and thus will need to use the respective debugger.

This is my first time looking at Meteor source so you might need to walk me through Barney style, crayons and all.   The strategy is to identify the call graph in the working case and then see where the divergence takes place in the failing case.

What's the correct way to instrument [livedata_connection.js](https://github.com/meteor/meteor/blob/devel/packages/ddp-client/livedata_connection.js) with log statements?

I tried ```console.log``` in:
```~/.meteor/packages/ddp-client/.1.3.2.3giexa++os+web.browser+web.cordova/web.browser/livedata_connection.js```

I get the following error on startup:
```
ppotoplyak@bigscreen:/pit/PixelUprise2$ meteor --version
Meteor 1.4.2.3
ppotoplyak@bigscreen:/pit/PixelUprise2$ meteor run
[[[[[ /pit/PixelUprise2 ]]]]]                 

=> Started proxy.                             
For compatibility, the PACKAGE_DIRS environment variable is deprecated and will be removed in a future Meteor release.
Developers should now use METEOR_PACKAGE_DIRS and Windows projects should now use a semi-colon (;) to separate paths.
=> Errors prevented startup:                  
   
   While computing assigned variables:
   /packages/ddp-client/livedata_connection.js: Unexpected token (1764:17)
   
=> Your application has errors. Waiting for file change.
=> Started MongoDB. 
```

I can dynamically add console.logs through the web browser console but it's a pain.  Again, the goal is to see the call graph in the working case.

**Comment by mitar** [comment]
I think the best way is to clone meteor repository itself, and then run your app using `meteor` script from there. So in your app just run `../path/to/meteor`. Then you can change `livedata_connection.js` file there (even while Meteor is running) and it will recompile the package and reload, with your console.log in there.

**Comment by ppotoplyak** [comment]
@mitar got ```console.log``` in ```livedata_connection.js```  working as you suggested, thanks.

Building the "successful" ```Meteor.call``` to callback call graph now.

**Comment by ppotoplyak** [comment]
@abernix good news, I am able to repro using [your reduced test case](https://github.com/abernix/meteor-issue-7868).

Server output (sorry about the text sizing as I wanted to capture the good->bad->good state transitions) : 
![dup](https://cloud.githubusercontent.com/assets/4952152/20925742/315f5ee4-bb7e-11e6-8908-2d3fc8366125.png)

You can see that the connection is good enough to keep calls coming in at a steady 3 sec interval.  This appears to be a core framework issue.

**Comment by abernix** [comment]
@ppotoplyak Unfortunately, I'm not sure how to help debug it as that same reproduction works fine for me.  So the question is: What is happening in your environment that isn't happening on mine?  Is high CPU your thinking right now?

**Comment by ppotoplyak** [comment]
@abernix if I had to speculate then it's a timing issue triggered by connection characteristics (could be total FUD).  My server is an ec2 instance, client is an A1429 iPhone/Safari/Latest Apple Bits, connection is T-Mobile pay as you go SIM showing a strong 4g connection.

I just tried again and lost callback 35 with < two minutes into the test:

```
<chop>
=> App running at: http://localhost:3000/
I20161207-18:24:50.548(0)? ->recordPing pingCounter : 0 
I20161207-18:24:53.075(0)? ->recordPing pingCounter : 1 
I20161207-18:24:56.153(0)? ->recordPing pingCounter : 2 
I20161207-18:24:59.154(0)? ->recordPing pingCounter : 3 
I20161207-18:25:02.114(0)? ->recordPing pingCounter : 4 
I20161207-18:25:05.125(0)? ->recordPing pingCounter : 5 
I20161207-18:25:08.124(0)? ->recordPing pingCounter : 6 
I20161207-18:25:11.174(0)? ->recordPing pingCounter : 7 
I20161207-18:25:14.134(0)? ->recordPing pingCounter : 8 
I20161207-18:25:17.183(0)? ->recordPing pingCounter : 9 
I20161207-18:25:20.144(0)? ->recordPing pingCounter : 10 
I20161207-18:25:23.144(0)? ->recordPing pingCounter : 11 
I20161207-18:25:26.195(0)? ->recordPing pingCounter : 12 
I20161207-18:25:29.154(0)? ->recordPing pingCounter : 13 
I20161207-18:25:32.163(0)? ->recordPing pingCounter : 14 
I20161207-18:25:35.164(0)? ->recordPing pingCounter : 15 
I20161207-18:25:38.174(0)? ->recordPing pingCounter : 16 
I20161207-18:25:41.214(0)? ->recordPing pingCounter : 17 
I20161207-18:25:44.184(0)? ->recordPing pingCounter : 18 
I20161207-18:25:47.184(0)? ->recordPing pingCounter : 19 
I20161207-18:25:50.314(0)? ->recordPing pingCounter : 20 
I20161207-18:25:53.194(0)? ->recordPing pingCounter : 21 
I20161207-18:25:56.194(0)? ->recordPing pingCounter : 22 
I20161207-18:25:59.244(0)? ->recordPing pingCounter : 23 
I20161207-18:26:02.244(0)? ->recordPing pingCounter : 24 
I20161207-18:26:05.214(0)? ->recordPing pingCounter : 25 
I20161207-18:26:08.254(0)? ->recordPing pingCounter : 26 
I20161207-18:26:11.224(0)? ->recordPing pingCounter : 27 
I20161207-18:26:14.224(0)? ->recordPing pingCounter : 28 
I20161207-18:26:17.264(0)? ->recordPing pingCounter : 29 
I20161207-18:26:20.314(0)? ->recordPing pingCounter : 30 
I20161207-18:26:23.244(0)? ->recordPing pingCounter : 31 
I20161207-18:26:26.284(0)? ->recordPing pingCounter : 32 
I20161207-18:26:29.294(0)? ->recordPing pingCounter : 33 
I20161207-18:26:32.904(0)? ->recordPing pingCounter : 34 
I20161207-18:26:35.697(0)? ->recordPing pingCounter : 34  DUP!
I20161207-18:26:38.272(0)? ->recordPing pingCounter : 36 
I20161207-18:26:41.273(0)? ->recordPing pingCounter : 37 
I20161207-18:26:44.282(0)? ->recordPing pingCounter : 38 
<chop>
```

Let me switch from cellular to wifi and see if things change.

When you try to repro are you hitting a local host or one in the cloud and are you on cellular or wifi?  I could be throwing mud at the wall, but my Galaxy S5 on Verizon has never ever hit this issue.

**Comment by ppotoplyak** [comment]
@abernix I apologize as I am muddling the water by not differentiating between lost vs delayed callbacks.  I think you suggested CPU usage as a factor in the delayed callbacks case?

For instance in my [previous post](https://github.com/meteor/meteor/issues/7868#issuecomment-265534313) you can see that callback 35 was delayed but not lost as the counter picked up at 36 after the DUP.   However, in my [first issue post ](https://github.com/meteor/meteor/issues/7868#issue-180984497) callbacks appear to be lost as the counter is not correct after the DUPs.

I forked and [updated your reduced test case](https://github.com/ppotoplyak/meteor-issue-7868) by moving the ```setTimeout``` inside the callback:
```
var pingCount = 0;
function heartbeat() {
  console.log('->recordPing pingCount: ' + pingCount);
  Meteor.call('recordPing', pingCount, function(err) {
    if(typeof err !== 'undefined') alert(err);
    console.log('<-recordPing pingCount: ' + pingCount);
    pingCount = pingCount + 1;
    setTimeout(heartbeat, 3000);
  });
}
heartbeat();
```

This way it will be obvious to see if a callback was lost.  Will let you know..

**Comment by pmgration** [comment]
I'm seeing something similar - callbacks for meteor method calls not firing sometimes and it seems to be caused by a "disconnection" between client and server (I quoted disconnection here because Meteor doesn't ever report that it is disconnected when logging Meteor.status().status, only ever reports `waiting` or `connecting`).
I've created a [video](https://www.youtube.com/watch?v=WjxNTlyYoJw) and made available on youtube to demonstrate what happens but to summarise what I'm experiencing:

- Meteor method calls work as expected while connected to the server
- If there's a short break in connection between client and server then when reconnecting meteor catches up by making the method calls that are queued up and client callbacks are fired
- But if the break in connection is long enough (say 30 seconds or more for me) then when reconnecting the queued method calls are made and received by server but the callback on the client isn't ever fired
- Subsequent calls to the method fail until browser is refreshed

Doesn't seem to be specific to one particular browser/version (I'm seeing the same with Chrome 54.0.2840.100 (64-bit) and Firefox 50.0.2, both on Linux).

Meteor version is: `1.4.2.3` (project has been upgraded several times through from `1.1.0.2`)

Here's the package list for the project:

```
accounts-password                  1.3.3
alanning:roles                     1.2.15
blaze-html-templates               1.0.5
coffeescript                       1.11.1_4*
iron:router                        1.0.13
jquery                             1.11.10
meteor-base                        1.0.4
mobile-experience                  1.0.4
momentjs:moment                    2.17.1
mongo                              1.1.14
mquandalle:jade                    0.4.9
mrt:accounts-admin-ui-bootstrap-3  0.2.7
pmgration:accounts-ldap            0.0.1+
session                            1.1.7
shell-server                       0.2.1
standard-minifier-css              1.3.2
standard-minifier-js               1.2.1
tracker                            1.1.1
twbs:bootstrap                     3.3.6
```

The project is quite big and I'm not able to make it available at the moment. I'm working on making a minimal version to reproduce but so far haven't been able to reproduce the problem. I'll try with the example posted earlier in this thread and see what happens but thought adding this *may* help with this issue.

**Comment by ppotoplyak** [comment]
@pmgration I am looking at the [video starting here](https://youtu.be/WjxNTlyYoJw?t=5m4s).  It looks like the system is in a state where client->server calls are being made as the server is acknowledging the invocations.  However, the callback is never fired on the client?

If so, this looks exactly like the state which I encountered minus the forced disconnects.  A refresh also fixed things in my case.  Hopefully I will have time over the break to look at this again.

Thanks for making the video.  I imagine you lost a good amount of time to this as I have.

**Comment by ppotoplyak** [comment]
@abernix just tried again using your [reduced test case](https://github.com/abernix/meteor-issue-7868) but lowering the call interval to one second which seems to exacerbate the issue.

235 seconds in, 102 callbacks were not _delayed_ but completely _lost_ :
```
<chop>
I20161224-04:05:32.629(0)? ->recordPing pingCounter : 231 
I20161224-04:05:33.639(0)? ->recordPing pingCounter : 232 
I20161224-04:05:34.640(0)? ->recordPing pingCounter : 233 
I20161224-04:05:35.640(0)? ->recordPing pingCounter : 234 
I20161224-04:05:36.649(0)? ->recordPing pingCounter : 235 
I20161224-04:05:37.649(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:38.660(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:39.660(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:40.659(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:41.659(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:42.659(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:43.660(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:44.710(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:45.709(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:46.709(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:47.670(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:48.679(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:49.679(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:50.679(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:51.680(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:52.679(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:53.679(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:54.691(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:55.689(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:56.689(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:57.690(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:58.689(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:05:59.700(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:00.699(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:01.700(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:02.700(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:03.699(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:04.669(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:05.709(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:06.739(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:07.749(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:08.740(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:09.749(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:10.719(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:11.719(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:12.719(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:13.719(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:14.719(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:15.749(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:16.729(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:17.729(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:18.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:19.729(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:20.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:21.779(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:22.739(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:23.739(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:24.739(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:25.739(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:26.750(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:27.750(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:28.749(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:29.749(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:30.759(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:31.760(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:32.761(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:33.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:34.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:35.770(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:36.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:37.810(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:38.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:39.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:40.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:41.769(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:42.809(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:43.779(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:44.780(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:45.779(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:46.779(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:47.779(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:48.790(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:49.749(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:50.789(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:51.789(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:52.790(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:53.789(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:54.800(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:55.801(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:56.800(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:57.799(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:58.799(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:06:59.810(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:00.809(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:01.810(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:02.810(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:03.809(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:04.849(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:05.819(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:06.819(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:07.819(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:08.820(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:09.819(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:10.819(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:11.829(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:12.830(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:13.830(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:14.829(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:15.829(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:16.839(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:17.839(0)? ->recordPing pingCounter : 235  DUP!
I20161224-04:07:18.839(0)? ->recordPing pingCounter : 236 
I20161224-04:07:19.840(0)? ->recordPing pingCounter : 237 
I20161224-04:07:20.839(0)? ->recordPing pingCounter : 238 
<chop>
```

Client is on an at&t sim, showing 3 of 5 4G bars, server is in the ec2 cloud.  Perhaps I am influenced by holiday revelry but this seems like a serious framework issue.

**Comment by ppotoplyak** [comment]
Unfortunately there is no differentiation of "completely lost callbacks" vs "long latency callbacks" in this issue which is my fault.  @pmgration I am curious to see if the callbacks would have fired given more time.

I just created #8194 to specifically address the long latency callback case.

**Comment by pmgration** [comment]
@ppotoplyak exactly, the method calls are making it to the server but the callback isn't ever fired on the client - I can make this reliably go wrong when forcing network disconnects in this way but a refresh always fixes it.
I first noticed this when running the app on a kiosk tablet (refresh and navigation bar etc disabled) and found that if I came back to it after a weekend for example the interface would appear to have frozen, of course this wasn't the case, after debugging and replicating on a desktop the was that the login callback wasn't ever being fired. So I have left it for long periods but haven't ever timed it or known exactly when it disconnected/stopped firing the callbacks, I'll try and do some more on this soon.

**Comment by ppotoplyak** [comment]
@pmgration if you can, add this [client code](https://raw.githubusercontent.com/ppotoplyak/meteor-issue-7868/master/client/main.js), and then this [server counterpart](https://raw.githubusercontent.com/ppotoplyak/meteor-issue-7868/master/server/main.js).  The code will log callback latencies to the server console.  It might shed some light or just create more confusion :)

**Comment by mitar** [comment]
One thing to try, maybe disable buffering on the client side by running `Meteor.connection._bufferedWritesInterval = 0;` Not sure if it should help, but maybe you find some differences.

**Comment by pmgration** [comment]
Unfortunately that didn't seem to make any difference for me @mitar. Haven't tried adding your debugging stuff yet @ppotoplyak.

**Comment by pmgration** [comment]
After quite a long time debugging yesterday it seems I've found the cause of the issue for me and it lies within the code of the project I'm working on...

One of the publications was written as follows (`server/publish.coffee`) which publishes some additional user information:

```coffeescript
Meteor.publish 'userData', ->
  if @userId
    Meteor.users.find @userId,
      fields:
        'services.ldap.cn': 1
        'services.ldap.displayName': 1
```

And on the client (client/app.coffee):

```coffeescript
Meteor.startup ->
  Meteor.subscribe 'userData'
  # ...etc...
```

Reading through the Meteor 1.4 documentation for publications it states:

"If a publish function does not return a cursor or array of cursors, it is assumed to be using the low-level added/changed/removed interface, and it must also call ready once the initial record set is complete."

The project started with Meteor 1.1 so I don't know if behaviour was different then or not but changing the publication (returning an empty array when the if condition isn't met) resolves the problem.

```coffeescript
Meteor.publish 'userData', ->
  if @userId
    Meteor.users.find @userId,
      fields:
        'services.ldap.cn': 1
        'services.ldap.displayName': 1
  []
```

What I don't yet understand is why it was only breaking when trying to make the method call when the server was offline and then came back online. Perhaps related to the order of subscribing/method calls  when the connection is re-established?

**Comment by andylash** [comment]
I was able to solve my problem by being more careful with promises in the method that was being called.  We had been returning before it was done, which was never correct but managed to work in the prior meteor releases.

**Comment by michalczaplinski** [comment]
I have had the same problem in Meteor 1.4.2.7.
When using `Meteor.call`, the callbacks are occasionally being dropped.
This also happens *only* on Safari. Using `Meteor.apply` instead, solves the issue..

What is more, when trying to debug the problem is Safari developer tools, the issue does not occur with the devloper tools *open*! :D

**Comment by mjmasn** [comment]
I think we've seen this issue intermittently - when it occurs we'll have a user complaining they can't log in (they just see the spinner forever), however our server logs show them logging in successfully.

Initially thought it was an issue with the user's browser because it worked fine for me but we're running behind a load balancer so I forced the connection to the same server as the user, and it turns out I couldn't log in either.

Haven't managed to reproduce the issue while devtools is open to see if the frames are coming through or not so I can't tell if it's a server or client issue. I'm guessing server-side because refreshing the page doesn't help and clearly the outgoing frames are working correctly.

Restarting the app on that server fixed the issue, but it has re-occurred 2 or 3 times over the last 6 months. Unfortunately we usually only notice if a customer complains because we use sticky-session cookies so we're only ever hitting a single server in our everyday use.

**Comment by pozylon** [comment]
Just had some strange issue right now with Meteor 1.5 where logout had not led to userId beeing recognized as changed, therefore Tracker did not recognize the new status, did not invalidate and UI did not refresh. No errors in server- or client side logs. After a hard browser-refresh, i got "Logged out by the server".

I restarted the node app and since everything works again as it should. It seems to me that incident was strongly related to this issue.

**Comment by mjmasn** [comment]
Still seeing this issue with Meteor 1.6. It's really frustrating because the app is totally fine, except no method callbacks get triggered on the client, so it's hard to check for. We normally realise or a customer informs us when login or logout is broken. Refreshing the page doesn't help even though this seems like a client-side issue. 

Seems to be [an issue that develops after a few days of continuous operation](http://www-users.math.umn.edu/~arnold/disasters/patriot.html), so if we haven't pushed an update for a while we start seeing issues. Maybe we should just start killing things regularly on a cronjob but that feels a bit heavy handed.

@pozylon are you running a load balancer? The issue we've had is that usually all but one server is working, so tracking down the issue is hard. If you're not using sticky sessions you might end up connecting to a working server the second time.

**Comment by pozylon** [comment]
The configuration was 2 instances behind nginx with sticky-session balancing.

It's going to be extremely hard to track that error down. I never had it show up again since my last post. If somebody can bring this further it it's maybe you @mjmasn. Would be really interesting how the frames look like that are sent over the ws if there are any at all, eliminating the client/server guessing so we can try to analyze some code.

**Comment by mjmasn** [comment]
@pozylon I had a memory of looking at the frames last time but I didn't post here and can't find any details so I'm probably imagining things, otherwise I'd have looked when the problem appeared again recently.

Like you say either the failure is on the server and we'll see no frames coming in, or it's a client issue and we'll receive the result frames but nothing happens with them, so at least it will be easy to narrow down, even if difficult to debug after that!

Will keep in mind for next time :) It's relatively easy to reproduce, just requires waiting a few days to a week without deploying/restarting/server reboot etc.

**Comment by jamesloper** [comment]
**So I happen to have a lot of info on this issue.** I have been talking to top developers and CTO's of various companies trying to diagnose this issue non-stop for the past 4 days.

Breaking down what happens behind the scenes...

  1. The `method` message is sent to the server
  2. The `result` message is sent to the client from the server
  3. The `updated` message is missing in action, it never ends up happening

According to Step 6 in the [Call Lifecycle](https://guide.meteor.com/methods.html#call-lifecycle) it explains why the callback doesn't fire.

We did figure out that `Meteor.disconnect(); Meteor.reconnect()` will make Meteor stop waiting for the `updated` message and just go ahead and fire off any pending callbacks. This has the effect of fixing the app. This is the bandaid that we used for two days.

I did eventually figure out that 1.4.4.3 doesn't have the problem but 1.5.2 does.

This is the course of action I propose:

- [ ] Insert a console.log right before the DDP server sends messages to the client and that will give us insight to make sure server is trying to send the `updated` message at all.
- [ ] See if this happens in an asteroid client! Asteroid may actually reveal something interesting 
 since the real app isn't bootstrapped, there shouldn't be any substantial logic running between Meteor server's results and updated replies (since there will be no publications because asteroid is run manually and you won't have subscribed to any publications*)

This should "split the world in two" and narrow down where to look. The only problem is, I want to help, but I am unable to test because I can't reproduce the issue anymore, after reverting prod to 1.4.4.3.

**Comment by jamesloper** [comment]
The bug has been solved by the latest version of `npm-mongo`!!!

Just run `meteor update npm-mongo`

In case anyone was wondering, the "updated" message doesn't get sent until oplog tailing "catches up", so it's possible that the bug which causes oplog tailing to stall could caused this problem.

From my limited understanding, it's now a little more clear HOW the npm-mongo fix worked. I used to, from the client run Meteor.disconnect() and Meteor.reconnect() and the callback would fire just fine, with the result. This is because its bypassing the oplog stall-- the callbacks just go ahead and fire when reconnecting, cuz you get a new Connection ID. I livedata can fetch the initial dataset but get stuck observing for live updates. I never checked that live updates actually stopped working all together, though I wish I caught that during testing. Of course that was hard considering I couldn't call any methods and change any data. Pretty sure this is resolved!

**Comment by hwillson** [comment]
Great news @jamesloper, thanks! This relates to the Mongo patch supplied by MDG, that was ultimately accepted by Mongo (for more details see the https://github.com/meteor/meteor/issues/8598 thread, or https://github.com/meteor/meteor/issues/8598#issuecomment-335977167 in particular). I'll close this for now - if anyone is still encountering this issue, please let us know.

**Comment by rubenpa** [comment]
I'm still having this issue, running on 1.6 and npm-mongo 2.2.33.  Seems to be the same as @mjmasn  -- Were you able to solve this?

**Comment by vsheyanov** [comment]
Same issue, 1.6, npm-mongo 2.2.33, Galaxy container, loginWithPassword stopped working when proper credentials were entered, but it did work when password was wrong (https://forums.meteor.com/t/meteor-issue-with-method-life-cycle-not-returning-ddp-msg-updated/39386/2?u=vsheyanov)
I did found this discussion after our app was restarted, so next time it occurs, I'll check websocket frames at least...

**Comment by rubenpa** [comment]
@vsheyanov were you able to find a fix for this?  No luck on my end yet.  Is it possible that this is galaxy related? I am able to login if I connect the database to my local server.

**Comment by vsheyanov** [comment]
@rubenpa this issue happened only once in our test environment, and we restarted Galaxy to fix it. So no actual fix on our side. But also we couldn't reproduce this issue - that's what bothers me...

**Comment by Pagebakers** [comment]
We are still experiencing these issues totally randomly, especially if we haven't deployed any updates for some time.

I just checked this live at one of our customers.

In the WS frames i can see the method being called, the document is saved and then getting a proper result. Then the callback is never called and the subscription that should be updated by the newly inserted document doesn't get updated and the app is 'stuck'

**Comment by jamesloper** [comment]
Just chiming in because I made a method one time that caused something very similar but wasn't related. The problem ended up being that I had used a promise that never resolved. It was supposed to process all items in an array at once but an edge case made the array empty, xausingy shoddy code to never resolved. That blocked the fiber on the connection, making it look like this bug was occuring when it was actually my code.

**Comment by Pagebakers** [comment]
We've been running on redis-oplog for a couple of months now in production and along with all other performance issues, this issue has disappeared as well.
